### PR TITLE
[BEAM-7356] Support generic field selections in Filter.

### PR DESCRIPTION
**Pull request by reuvenlax**
This allows filtering on nested fields

### CODE REVIEW

**Comment by reuvenlax** [comment]
R: @robinyqiu

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java [hunk 1]
```diff
@@ -84,95 +87,128 @@
...
-    private final Map<List<Integer>, SerializableFunction<Row, Boolean>> fieldIdsFilters =
-        Maps.newHashMap();
+    @AutoValue
+    abstract static class FilterDescription<FieldT> implements Serializable {
+      abstract FieldAccessDescriptor getFieldAccessDescriptor();
+
+      abstract SerializableFunction<FieldT, Boolean> getPredicate();
+
+      @Nullable
+      abstract Schema getOutputSchema();
```
Comment (robinyqiu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java]: Nit: I found the name `outputSchema` a bit confusing, maybe something like `selectedSchema` or 'accessedSchema' is a better name?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java [hunk 2]
```diff
@@ -84,95 +87,128 @@
...
+
+      abstract boolean getUnbox();
```
Comment (robinyqiu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java]: Nit: same as above, maybe `selectedSingleField` or `accessedSingleField`?

**Comment by reuvenlax** [comment]
"generic" in the PR title meant generic field access strings. This PR
allows passing "x.y.z" into filter to filter on subfields, while the
previous Filter using did not allow this.

I made one change to actual Java generics, to make the transform better
able to infer types from lambdas. Previously you had to cast the type
inside your lambda, and now that is not necessary.

Good point on confusing names. I'll try and fix those.

**Comment by robinyqiu** [comment]
Thanks for the explanation! That answers my question.

Another quick question:
I think we can't use `java.util.function.Predicate<FieldT>` to replace `SerializableFunction<FieldT, Boolean>` defined in Beam because `Predicate` doesn't implement `Serializable`, right? Then maybe adding something like `SerializablePredicate` to the class of functional interfaces defined by Beam will be helpful.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java (hunk 2 above)
Comment (robinyqiu) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java]: Thank you for the changes! LGTM.

**Pull request merged**